src/VisualStudio/Core/Def/Telemetry/TelemetryLogger.cs

```diff
@@ -5,29 +5,40 @@
 using System;
 using System.Collections.Concurrent;
 using System.Collections.Generic;
-using System.Diagnostics;
 using System.Linq;
 using System.Threading;
-using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Internal.Log;
-using Microsoft.CodeAnalysis.Options;
+using Microsoft.VisualStudio.LanguageServices.Telemetry;
 using Microsoft.VisualStudio.Telemetry;
 using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.Telemetry
 {
     internal abstract class TelemetryLogger : ILogger
     {
-        private sealed class Implementation : TelemetryLogger
+        private sealed class Implementation : TelemetryLogger, IDisposable
         {
             private readonly TelemetrySession _session;
+            private TelemetryLogProvider? _telemetryLogProvider;
 
             public Implementation(TelemetrySession session, bool logDelta)
             {
                 _session = session;
                 LogDelta = logDelta;
             }
 
+            public void Dispose()
+            {
+                _telemetryLogProvider?.Dispose();
+            }
+
+            public void Initialize()
+            {
+                // Two stage initialization as TelemetryLogProvider.Create needs access to
+                //  the ILogger that this class implements.
+                _telemetryLogProvider = TelemetryLogProvider.Create(_session, this);
```

Comment:
Assert `_telemetryLogProvider == null` here?

Reply:
The initialize method isn't really intended to be public. Revamped a bit in commit 3 such that creation and setting of this member occur from the same method.